File: typescript/src/day05/Almanac.ts
import { ensure } from '../utils'
import AlmanacMap from './AlmanacMap'


export class Almanac {

  private seeds = [] as [number, number][]
  private maps = new Map<string, AlmanacMap>()

  constructor() {
  }

  getMapFor(source: string): AlmanacMap {
    return ensure(this.maps.get(source))
  }

  addSeeds(seeds: [number, number][]): Almanac {
    this.seeds.push(...seeds)
    return this
  }

  getSeeds(): [number, number][] {
    return this.seeds
  }

  addMap(map: AlmanacMap): Almanac {
    this.maps.set(map.source, map)

    return this
  }

  static parse(lines: string[], seedAdapter: (seeds: number[]) => [number, number][]): Almanac {
    const ret = new Almanac()

    const seeds = lines[0]
      ?.split(":")[1]
      ?.split(" ")
      .map(_ => _.trim())
      .filter(_ => _ != '')
      .map(_ => Number.parseInt(_)) ?? []

    ret.addSeeds(seedAdapter(seeds))

    const buffer = [] as string[]

    for (let i = 1; i < lines.length; i++) {
      const line = lines[i].trim()
      if (line != '') {
        buffer.push(line)
      } else if (buffer.length > 0) {
        ret.addMap(AlmanacMap.parse(buffer))
        buffer.length = 0
      }
    }

    if (buffer.length > 0)
      ret.addMap(AlmanacMap.parse(buffer))

    return ret
  }

}

export default Almanac

File: AoCKotlin/src/main/kotlin/day05/Almanac.kt
package me.brasten.day05

enum class Category {
    SEED,
    SOIL,
    FERTILIZER,
    WATER,
    LIGHT,
    TEMPERATURE,
    HUMIDITY
}

class AlmanacMap {

}

data class SeedRange(
    val start: Int,
    val length: Int,
)

class Almanac {
    private val categories: MutableMap<Category, AlmanacMap> = mutableMapOf()
    private val seedRanges: MutableList<SeedRange> = mutableListOf()

    fun addSeeds(seeds: List<SeedRange>) {
        seedRanges.addAll(seeds)
    }

    companion object {
        fun parse(lines: List<String>): Almanac {
            val retVal = Almanac()
            val buffer = mutableListOf<String>()

            for (line in lines) {
                if (line.startsWith("seeds:")) {
                    val seeds = line.split(":")
                        .last()
                        .split(" ")
                        .map { it.trim() }
                        .filter { it !== "" }
                        .map { it.toInt() }
                        .chunked(2)
                        .map { SeedRange(it[0], it[1]) }

                    retVal.addSeeds(seeds)
                } else if (line.trim() == "") {
                    if (buffer.isNotEmpty()) {

                    }
                } else {

                }
            }

//            println("seeds -> ${seeds}")

//            println("next line -> ${lineIter.take(3).toList()}")

            return retVal
        }
    }
}